mongo/integration/client_side_encryption_prose_test.go

```diff
@@ -1088,6 +1163,192 @@ func TestClientSideEncryptionProse(t *testing.T) {
+				"tenantId":                 azureTenantID,
+				"clientId":                 azureClientID,
+				"clientSecret":             azureClientSecret,
+				"identityPlatformEndpoint": "127.0.0.1:8001",
+			},
+			"gcp": {
+				"email":      gcpEmail,
+				"privateKey": gcpPrivateKey,
+				"endpoint":   "127.0.0.1:8001",
+			},
+			"kmip": {
+				"endpoint": "127.0.0.1:8001",
+			},
+		}
+
+		// create valid Client Encryption options without TLS options
+		validClientEncryptionOptionsWithoutTLS := options.ClientEncryption().
+			SetKmsProviders(validKmsProviders).
+			SetKeyVaultNamespace(kvNamespace)
+
+		// make TLS opts containing client certificate and CA file
+		tlsOptsMap := make(map[string]map[string]interface{})
+		clientAndCATlsMap := map[string]interface{}{
+			"tlsCertificateKeyFile": sslClientCertificateKeyFile,
+			"tlsCAFile":             sslCertificateAuthorityFile,
+		}
+		tlsOptsMap["kmip"] = clientAndCATlsMap
+
+		// create valid Client Encryption options and set valid TLS options
+		validClientEncryptionOptionsWithTLS := options.ClientEncryption().
+			SetKmsProviders(validKmsProviders).
+			SetKeyVaultNamespace(kvNamespace)
+
+		_, err := validClientEncryptionOptionsWithTLS.SetTLSConfig(tlsOptsMap)
+		assert.Nil(mt, err, "expected no error, got error: %v", err)
+
+		// make TLS opts containing only CA file
+		caTlsMap := map[string]interface{}{
+			"tlsCAFile": sslCertificateAuthorityFile,
+		}
+		tlsOptsMap["kmip"] = caTlsMap
+
+		// create invalid Client Encryption options with expired credentials
+		expiredClientEncryptionOptions := options.ClientEncryption().
+			SetKmsProviders(expiredKmsProviders).
+			SetKeyVaultNamespace(kvNamespace)
+
+		_, err = expiredClientEncryptionOptions.SetTLSConfig(tlsOptsMap)
+		assert.Nil(mt, err, "expected no error, got error: %v", err)
+
+		// create invalid Client Encryption options with invalid hostnames
+		invalidHostnameClientEncryptionOptions := options.ClientEncryption().
+			SetKmsProviders(invalidKmsProviders).
+			SetKeyVaultNamespace(kvNamespace)
+
+		_, err = invalidHostnameClientEncryptionOptions.SetTLSConfig(tlsOptsMap)
+		assert.Nil(mt, err, "expected no error, got error: %v", err)
+
+		awsMasterKey := map[string]interface{}{
+			"region":   "us-east-1",
+			"key":      "arn:aws:kms:us-east-1:579766882180:key/89fcc2c4-08b0-4bd9-9f25-e30687b580d0",
+			"endpoint": "127.0.0.1:8002",
+		}
+		azureMasterKey := map[string]interface{}{
+			"keyVaultEndpoint": "doesnotexist.local",
+			"keyName":          "foo",
+		}
+		gcpMasterKey := map[string]interface{}{
+			"projectId": "foo",
+			"location":  "bar",
+			"keyRing":   "baz",
+			"keyName":   "foo",
+		}
+		kmipMasterKey := map[string]interface{}{}
+
+		testCases := []struct {
+			name                 string
+			masterKey            interface{}
+			expiredError         string
+			invalidHostnameError string
+		}{
+			{"aws", awsMasterKey, "certificate signed by unknown authority", "certificate signed by unknown authority"},
```

Comment:
Wanted to call attention to this because all the broken implementations for the AWS test case seem to raise the same "x509: certificate signed by unknown authority" error. According to the [CSFLE spec](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#case-1-aws), error messages should more specifically describe the reason for a TLS handshake failure.

I'm not sure if this has something to do with how AWS handles failed TLS handshakes but thought I'd point it out in case error messages should be consistent across KMS providers.

Reply:
Oof, that was my mistake that was addressed on a later commit on DRIVERS-1353 https://github.com/mongodb/specifications/commit/676a4623379b211d2e29a81c73ab5bf5d47ab708. The `tlsCAFile` should be configured for all KMS providers.